cookbooks/finite_strain/finite_strain.cc

```diff
@@ -0,0 +1,114 @@
+#include <aspect/material_model/simple.h>
+#include <aspect/simulator_access.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    template <int dim>
+    class FiniteStrain : public MaterialModel::Simple<dim>
+    {
+      public:
+
+        virtual void evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+                              MaterialModel::MaterialModelOutputs<dim> &out) const;
+        virtual void parse_parameters(ParameterHandler &prm);
+    };
+
+  }
+}
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+
+    template <int dim>
+    void
+    FiniteStrain<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      Simple<dim>::evaluate(in, out);
+
+      // we need the velocity gradiens for the finite strain (they are not included in material model inputs)
+      // so we get them from the finite element
+      if(in.cell && this->get_timestep_number() > 0)
+      {
+        const QGauss<dim> quadrature_formula (this->get_fe().base_element(this->introspection().base_elements.velocities).degree+1);
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_values   |
+                                 update_gradients |
+                                 update_quadrature_points |
+                                 update_JxW_values);
+
+        std::vector<Tensor<2,dim> > velocity_gradients (quadrature_formula.size(), Tensor<2,dim>());
+
+        fe_values.reinit (*in.cell);
+        fe_values[this->introspection().extractors.velocities].get_function_gradients (this->get_solution(),
+                                                                                       velocity_gradients);
+
+        // Assign the strain rate components to the compositional fields reaction terms.
+        // We also have to rotate the accumulated strain from all the previous time steps with
+        // the rotation part of the flow field of the current time step.
+        // If there are too many fields, we simply fill only the first fields with the
+        // existing strain rate tensor components.
+        for (unsigned int q=0; q < in.position.size(); ++q)
+        {
+          // rotation tensor = asymmetric part of the velocity gradients tensor + unit tensor
+          const Tensor<2,dim> rotation = (velocity_gradients[q] - in.strain_rate[q]) * this->get_timestep()
+                                         + unit_symmetric_tensor<dim>();
+
+          Tensor<2,dim> strain;
```

Comment:
Oh, I already updated this to `accumulated_strain` (it only shows the old version because this is what Wolfgang's comment was referring to). But if you click below on the commit  `add a cookbook that describes tracking finite strain` or the `Files changed` tab you will see the most recent version (just without the comments).